nav2_map_server/README.md

```diff
@@ -1,59 +1,133 @@
+
+`MapIO 3D` - is a 3D analogue of `MapIO 2D`: map input-output library for `PointClouds` and deals especially with `pcd`. For more details about `pcd` file format please refer to [pcl website](https://pcl.readthedocs.io/projects/tutorials/en/latest/pcd_file_format.html#pcd-file-format).
 
-The command line for the map server executable is slightly different that it was with ROS1.
-With ROS1, one invoked the map server and passing the map YAML filename, like this:
+---
+**Note:** In the present implementation the `MapIO 2D` and  `MapIO 3D` libraries reflects their respective *specializations* conpatible with `MapServer` and `MapSaver`. 
 
+For `MapIO 2D`:
+```c++
+// map_2d/map_server_2d.hpp
+// map_2d/map_server_2d.cpp
+template<>
+class MapServer<nav_msgs::msg::OccupancyGrid> : public nav2_util::LifecycleNode {...};
+
+// map_2d/map_saver_2d.hpp
+// map_2d/map_saver_2d.cpp
+template<>
+class MapSaver<nav_msgs::msg::OccupancyGrid> : public nav2_util::LifecycleNode {...};
 ```
-$ map_server map.yaml
+
+For `MapIO 3D`:
+```c++
+// map_3d/map_server_3d.hpp
+// map_3d/map_server_3d.cpp
+template<>
+class MapServer<sensor_msgs::msg::PointCloud2> : public nav2_util::LifecycleNode {...};
+
+// map_3d/map_saver_3d.hpp
+// map_3d/map_saver_3d.cpp
+template<>
+class MapSaver<sensor_msgs::msg::PointCloud2> : public nav2_util::LifecycleNode {...};
 ```
 
-Where the YAML file specified contained the various map metadata, such as:
+---
+Listed here all the main vending points need to be included, to get the full functionality of the presently implemented library components.
+
+### Library interfaces list
+| Basic Components<br />nav2_map_server/ | Functionality | Namespace | Contains
+| :-----------------: | :-----------------: | :------------------: | :------------: |
+| map_server.hpp | Template Map Server | nav2_map_server | `MapServer<OccupancyGrid>()`<br /> `MapServer<Pointcloud2>()`
+| map_saver.hpp | Template Map Saver | nav2_map_server | `MapSaver<OccupancyGrid>()`<br /> `MapSaver<Pointcloud2>()`
+| map_io_2d.hpp | 2D Map IO | nav2_map_server::map_2d | Functions for OccupancyGrid I/O operations
+| map_io_3d.hpp | 3D Map IO | nav2_map_server::map_3d | Functions for PointCloud2 I/O operations
+
 
+---
+### CLI-usage
+
+The existing system of Nav2 Map Server provides many executables to get a hassle free services launch and handling. Detailed list and functionalities are given below:
+
+| Executables | Implementation | Functionality | Type | Map Type
+| :-------------: | :-----------------: | :--------------: | :----: | :--- |
+| map_server_2d | main_2d.cpp | Map Server | LifcyleNode | OccupancyGrid
+| map_server_3d | main_3d.cpp | Map Server | LifcyleNode | PointCloud2
+| map_saver_server_2d | main_server_2d.cpp | Map Saver | LifcyleNode | OccupancyGrid
+| map_saver_server_3d | main_server_3d.cpp | Map Saver | LifcyleNode | PointCloud2
+| map_saver_cli | main_cli.cpp | Map Saver CLI | CLI | Detects based<br /> on input parameters
+  
```

Comment:
Here is a mix of LifecycleNodes that could be used from CLI, Services (with no direct CLI usage designed) and one directly used CLI-executable. To avoid a confusion, this part (starting from "The existing system ..." ending with the table above) should be moved to "Architecture" section (in my opinion, placed in the end of it).

Reply:
I differ from this, as this list includes all the executables, present in the `nav2_map_server`, also this list isn't representative of architecture, rather it is more like all the possible executables. If lifecycle manager is a problem for some executable, I have added a demonstration, of how to use a custom node to change the state.

Reply:
No need to demonstrate how to change lifecycle node state (this is known thing).
The question is not here. Initially documentation was grouped to 1. CLI-executables and 2. Service nodes.
This table contains both of them, but for some reasons belonging to CLI. However, not all executables (even if they can be run from the command line) are belong to "CLI" term in this documentation. Let's clarify:
* CLI - is that designed to be called through command line interface only once. In this list are: `map_saver_cli` and `map_server_2d/3d` (the last being called through once activating LifecycleNode's state)
* Service server - is that designed to operate as a continuslty running ROS2 node, the server handling incoming service requests. In this list are: again `map_server_2d/3d` (it can act both as CLI and server) and `map_saver_server_2d/3d` executables.

This table have mixed list of all executables, so I still insisting to move it out of CLI section (if not to the architecture, that to somewhere else).

Reply:
Dumped this list in a separate section to list down all the implemented executables. irrespective of their use, the list tends to provide the information of all the executables in one place.

Reply:
This was moved to "Existing Executables" section. Resolving.